README.md

```diff
@@ -1,86 +1,100 @@
+* an address space and 
+* a context in which communication can happen.
+* TODO: quality of service
+
+A Peer Group could be something like scalanet's `UDPPeerGroup` where the addresses are IP:port combos and the set of 
+peers allowed to communicate is basically anybody on the IP network in question (the internet, an office network, etc).
+Equally, a Peer Group could be something like an Ethereum network where addresses are public keys and the peers
+are anybody who talks the RLPx protocol. Equally, a peer group could be an integration test with the address space {Alice, Bob, Charlie}
+and the peers are all in the same JVM. Scalanet will not limit you in this regard. 
+
+Peer groups can implement arbitrary enrolment and encryption schemes, so are suitable for implementing secure messaging overlays.
+Typically, on the internet, limiting the context of communication (aka _zoning_) is performed by firewalls. The idea
+is so ubiquitous that it may not have struck you that this is a hack. Peer groups are designed to support more elegant
+solutions, generally using cryptography instead of firewall provisioning.
+ 
+### Structure of the library
+Here is a picture of the structure of the library for a few sample applications:
+![](doc-resources/sample-configurations.png)
+
+### Getting started
+The easiest way to get started is to send and receive data over TCP, using the library just like netty. Have a look at
+the [TCPPeerGroupSpec](src/io/iohk/scalanet/test/peergroup/TCPPeerGroupSpec.scala) test case or the following code. 
+
+```scala
+// import some peer group classes
+import io.iohk.scalanet.peergroup._
+
+// message sending can be controlled using either
+// monix, cat-effect, cats EitherT or scala Futures
+// depending on your taste.
+import io.iohk.scalanet.peergroup.future._
+import scala.concurrent.Future
+
+import java.net.InetSocketAddress
+import java.nio.ByteBuffer
+
+val config = TCPPeerGroup.Config(new InetSocketAddress(???))
+
+val tcp = TCPPeerGroup.createOrThrow(config)
+
+// send a message
+val messageF: Future[Unit] = tcp.sendMessage(new InetSocketAddress("example.com", 80), ByteBuffer.wrap("Hello!".getBytes))
+
+// receive messages
+tcp.messageStream.foreach((b: ByteBuffer) => ())
+ 
+``` 
+
+# Contributing
+
+### Branches
 
 Two main branches are maintained: `develop` and `master`. 
-`master` contains the latest version of the code that was tested end-to-end. 
-`develop` contains the latest version of the code that runs all the tests (unit and integration tests). 
-
+`master` contains the latest stable version of the library. 
+`develop` is the place you want to merge to if submitting PRs.
 
-## Working with the codebase
+### Building the codebase
 
-To build the codebase, we use [bazel](https://bazel.build/).
+To build the codebase, we use [bazel](https://bazel.build/). Assuming you have bazel installed correctly, you can build
+the codebase with
+```bash
+bazel test //...
```

Comment:
`bazel build //...`, otherwise, change the previous line to `you can run the codebase tests with`